Framework/Muon/src/MuonPairingAsymmetry.cpp

```diff
@@ -0,0 +1,476 @@
+  validatePeriods(inputWS, errors);
+}
+
+void MuonPairingAsymmetry::validateGroupsWorkspaces(
+    std::map<std::string, std::string> &errors) {
+  Workspace_sptr ws1 = this->getProperty("InputWorkspace1");
+  Workspace_sptr ws2 = this->getProperty("InputWorkspace2");
+  if (isMultiPeriod(ws1) && !isMultiPeriod(ws2)) {
+    errors["InputWorkspace1"] =
+        "InputWorkspace2 should be multi period to match InputWorkspace1";
+  }
+  if (isMultiPeriod(ws2) && !isMultiPeriod(ws1)) {
+    errors["InputWorkspace2"] =
+        "InputWorkspace1 should be multi period to match InputWorkspace2";
+  }
+  if (!checkConsistentPeriods(ws1, ws2)) {
+    errors["InputWorkspace1"] = "InputWorkspace1 and InputWorkspace2 have "
+                                "inconsistent numbers of periods.";
+  }
+  if (isMultiPeriod(ws1) && isMultiPeriod(ws2)) {
+    validatePeriods(boost::dynamic_pointer_cast<WorkspaceGroup>(ws1), errors);
+    validatePeriods(boost::dynamic_pointer_cast<WorkspaceGroup>(ws2), errors);
+  }
+}
+
+void MuonPairingAsymmetry::exec() {
+
+  MatrixWorkspace_sptr outWS;
+
+  if (getProperty("SpecifyGroupsManually")) {
+    outWS = execSpecifyGroupsManually();
+  } else {
+    outWS = execGroupWorkspaceInput();
+  }
+
+  setPairAsymmetrySampleLogs(outWS);
+  setProperty("OutputWorkspace", outWS);
+}
+
+MatrixWorkspace_sptr MuonPairingAsymmetry::execGroupWorkspaceInput() {
+
+  // Get the input workspace into a useful form
+  Workspace_sptr tmpWS1 = getProperty("InputWorkspace1");
+  Workspace_sptr tmpWS2 = getProperty("InputWorkspace2");
+  WorkspaceGroup_sptr ws1 = workspaceToWorkspaceGroup(tmpWS1);
+  WorkspaceGroup_sptr ws2 = workspaceToWorkspaceGroup(tmpWS2);
+  WorkspaceGroup_sptr groupedPeriods = boost::make_shared<WorkspaceGroup>();
+  for (int i = 0; i < countPeriods(ws1); i++) {
+    groupedPeriods->addWorkspace(
+        appendSpectra(getWorkspace(ws1, i), getWorkspace(ws2, i)));
+  }
+
+  // Do the asymmetry calculation
+  const double alpha = static_cast<double>(getProperty("Alpha"));
+  std::vector<int> summedPeriods = getProperty("SummedPeriods");
+  std::vector<int> subtractedPeriods = getProperty("SubtractedPeriods");
+  return calcAsymmetryWithSummedAndSubtractedPeriods(
+      summedPeriods, subtractedPeriods, groupedPeriods, alpha);
+}
+
+MatrixWorkspace_sptr MuonPairingAsymmetry::execSpecifyGroupsManually() {
+
+  WorkspaceGroup_sptr inputWS = getProperty("InputWorkspace");
+  auto groupedPeriods = createGroupWorkspace(inputWS);
+
+  // Do the asymmetry calculation
+  std::vector<int> summedPeriods = getProperty("SummedPeriods");
+  std::vector<int> subtractedPeriods = getProperty("SubtractedPeriods");
+  const double alpha = static_cast<double>(getProperty("Alpha"));
+
+  return calcAsymmetryWithSummedAndSubtractedPeriods(
+      summedPeriods, subtractedPeriods, groupedPeriods, alpha);
+}
+
+MatrixWorkspace_sptr
+MuonPairingAsymmetry::calcAsymmetryWithSummedAndSubtractedPeriods(
```

Comment:
we should change this methods name. The asymmetry is not calculated in this code and is instead estimated (we need more information to do a calculated asymmetry)